tsl/src/continuous_aggs/planner.c

```diff
@@ -355,3 +355,101 @@ constify_cagg_watermark(Query *parse)
 	if (context.valid_query)
 		replace_watermark_with_const(&context);
 }
+
+/*
+ * Push down ORDER BY and LIMIT into subqueries of UNION for realtime
+ * continuous aggregates when sorting by time.
+ */
+void
+cagg_sort_pushdown(Query *parse, int *cursor_opts)
+{
+	ListCell *lc;
+
+	/* Nothing to do if we have no valid sort clause */
+	if (list_length(parse->rtable) != 1 || list_length(parse->sortClause) != 1 ||
+		!OidIsValid(linitial_node(SortGroupClause, parse->sortClause)->sortop))
+		return;
+
+	Cache *cache = ts_hypertable_cache_pin();
+
+	foreach (lc, parse->rtable)
+	{
+		RangeTblEntry *rte = lfirst(lc);
+
+		/*
+		 * Realtime cagg view will have 2 rtable entries, one for the materialized data and one for
+		 * the not yet materialized data.
+		 */
+		if (rte->rtekind != RTE_SUBQUERY || rte->relkind != RELKIND_VIEW ||
+			list_length(rte->subquery->rtable) != 2)
+			continue;
+
+		ContinuousAgg *cagg = ts_continuous_agg_find_by_relid(rte->relid);
+
+		/*
+		 * This optimization only applies to realtime caggs.
+		 */
+		if (!cagg || !cagg->data.finalized || cagg->data.materialized_only)
+			continue;
+
+		Hypertable *ht = ts_hypertable_cache_get_entry_by_id(cache, cagg->data.mat_hypertable_id);
+		Dimension const *dim = hyperspace_get_open_dimension(ht->space, 0);
+
+		/* We should only encounter hypertables with an open dimension */
+		if (!dim)
+			continue;
+
+		SortGroupClause *sort = linitial_node(SortGroupClause, parse->sortClause);
+		TargetEntry *tle = get_sortgroupref_tle(sort->tleSortGroupRef, parse->targetList);
+
+		/*
+		 * We only pushdown ORDER BY when it's single column
+		 * ORDER BY on the time column.
+		 */
+		AttrNumber time_col = dim->column_attno;
+		if (!IsA(tle->expr, Var) || castNode(Var, tle->expr)->varattno != time_col)
+			continue;
```

Comment:
Why we can't pushdown ORDER BY clauses that include the primary dimension plus other columns?